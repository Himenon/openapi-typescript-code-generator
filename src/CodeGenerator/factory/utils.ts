import { EOL } from "os";

export interface Comment {
  hasTrailingNewLine: boolean;
  value: string;
}

export const isOnlyAlphabetText = (text: string): boolean => {
  return new RegExp(/^[a-zA-Z]+$/).test(text);
};

export const escapeIdentiferText = (text: string): string => {
  return text.replace(/-/g, "_");
};

export const generateComment = (comment: string, deprecated?: boolean): Comment => {
  const splitComments = deprecated ? ["@deprecated"].concat(comment.split(EOL)) : comment.split(EOL);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  if (comments.length === 1) {
    return {
      hasTrailingNewLine: true,
      value: "* " + comments.join("") + " ",
    };
  }
  return {
    hasTrailingNewLine: true,
    value: "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ",
  };
};
