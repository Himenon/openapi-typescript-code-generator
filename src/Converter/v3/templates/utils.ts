import ts from "typescript";

import { DevelopmentError } from "../../../Exception";
import { Factory } from "../../../TypeScriptCodeGenerator";

export interface StringItem {
  type: "string";
  value: string;
}

export interface ExpressionItem {
  type: "property";
  value: ts.Expression;
}

export type Item = StringItem | ExpressionItem;

export type Params$TemplateExpression = Item[];

export const isAlphabetOnlyText = (text: string): boolean => {
  return /^[A-Za-z\s]+$/.test(text);
};

const getTemplateSpan = (
  factory: Factory.Type,
  currentIndex: number,
  nextIndex: number,
  lastIndex: number,
  currentItem: ExpressionItem,
  nextItem: Item | undefined,
): ts.TemplateSpan[] => {
  // == last
  if (!nextItem) {
    return [
      factory.TemplateSpan.create({
        expression: currentItem.value,
        literal: factory.TemplateTail.create({
          text: "",
        }),
      }),
    ];
  }

  if (nextItem.type === "property") {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateMiddle.create({
            text: "",
          }),
        }),
      ];
    }
  } else {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: nextItem.value,
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateMiddle.create({
            text: nextItem.value,
          }),
        }),
      ];
    }
  }
};

/**
 *
 * ``
 * `a`
 * `${b}`
 * `a${b}`
 * `${a}${b}`
 * `${a}b${c}`
 * ``
 */
export const generateTemplateExpression = (factory: Factory.Type, list: Params$TemplateExpression): ts.Expression => {
  if (list.length === 0) {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: "",
    });
  }
  const templateHead = factory.TemplateHead.create({
    text: list[0].type === "property" ? "" : list[0].value,
  });

  const spanList = list.splice(1, list.length);
  if (spanList.length === 0 && list[0].type === "string") {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: list[0].value,
    });
  }
  const lastIndex = spanList.length - 1;
  const restValue = lastIndex % 2;
  let templateSpans: ts.TemplateSpan[] = [];
  for (let i = 0; i <= (lastIndex - restValue) / 2; i++) {
    if (spanList.length === 0) {
      continue;
    }
    const currentIndex = 2 * i;
    const nextIndex = 2 * i + 1;
    const currentItem = spanList[currentIndex];
    const nextItem = spanList[nextIndex];
    if (currentItem.type === "string") {
      throw new DevelopmentError("Logic Error");
    }
    templateSpans = templateSpans.concat(getTemplateSpan(factory, currentIndex, nextIndex, lastIndex, currentItem, nextItem));
  }
  return factory.TemplateExpression.create({
    head: templateHead,
    templateSpans: templateSpans,
  });
};

export const generateVariableIdentifier = (
  factory: Factory.Type,
  name: string,
): ts.Identifier | ts.PropertyAccessExpression | ts.ElementAccessExpression => {
  if (name.startsWith("/")) {
    throw new Error("can't start '/'. name=" + name);
  }
  const list = name.split(".");
  if (list.length === 1) {
    return factory.Identifier.create({
      name: name,
    });
  }
  const [n1, n2, ...rest] = list;
  const first = factory.PropertyAccessExpression.create({
    expression: n1,
    name: n2,
  });

  return rest.reduce<ts.PropertyAccessExpression | ts.ElementAccessExpression>((previous, current: string) => {
    if (isAlphabetOnlyText(current)) {
      return factory.PropertyAccessExpression.create({
        expression: previous,
        name: current,
      });
    }
    return factory.ElementAccessExpression.create({
      expression: previous,
      index: current,
    });
  }, first);
};

export interface LiteralExpressionObject {
  [key: string]: { type: "string" | "variable"; value: string };
}

export const generateObjectLiteralExpression = (
  factory: Factory.Type,
  obj: LiteralExpressionObject,
  extraProperties: ts.PropertyAssignment[] = [],
): ts.ObjectLiteralExpression => {
  const properties = Object.entries(obj).map(([key, item]) => {
    const initializer =
      item.type === "variable" ? generateVariableIdentifier(factory, item.value) : factory.StringLiteral.create({ text: item.value });
    return factory.PropertyAssignment.create({
      name: isAlphabetOnlyText(key) ? key : `"${key}"`, // TODO escape _ / . ...etc
      initializer,
    });
  });

  return factory.ObjectLiteralExpression.create({
    properties: extraProperties.concat(properties),
    multiLine: true,
  });
};
