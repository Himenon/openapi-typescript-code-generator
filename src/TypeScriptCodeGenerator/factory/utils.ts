import { EOL } from "os";

export const generateComment = (comment: string, deprecated?: boolean): string => {
  const splitComments = deprecated ? ["@deprecated"].concat(comment.split(EOL)) : comment.split(EOL);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  return "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ";
};
