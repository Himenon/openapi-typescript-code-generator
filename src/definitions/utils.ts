import { JSONSchema7 } from "json-schema";

export type JSONSchema = JSONSchema7;

export type MapLike<K extends string, T> = {
  [key in K]: T;
};
