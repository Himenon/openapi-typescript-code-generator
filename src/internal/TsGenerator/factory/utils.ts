import { EOL } from "os";

export interface Comment {
  hasTrailingNewLine: boolean;
  value: string;
}

export const isOnlyAlphabetText = (text: string): boolean => {
  return new RegExp(/^[a-zA-Z]+$/).test(text);
};

export const escapeIdentiferText = (text: string): string => {
  return text.replace(/-/g, "_");
};

export const generateComment = (comment: string, deprecated?: boolean): Comment => {
  const excapedComment = comment
    .replace(/\*\//, "\\*\\\\/") // */  -> \*\/
    .replace(/\/\*/, "/\\\\*") // /*  -> \/\*
    .replace(/\*\/\*/, "\\*\\/\\*"); // */* -> \*\/\*
  const splitComments = deprecated ? ["@deprecated"].concat(excapedComment.split(/\r?\n/)) : excapedComment.split(/\r?\n/);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  if (comments.length === 1) {
    return {
      hasTrailingNewLine: true,
      value: "* " + comments.join("") + " ",
    };
  }
  return {
    hasTrailingNewLine: true,
    value: "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ",
  };
};
