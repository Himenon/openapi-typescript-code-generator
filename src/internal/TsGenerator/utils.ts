import ts from "typescript";

import type { CodeGenerator } from "../../types";

export const stringToStatements = (code: string): ts.Statement[] => {
  const source = ts.createSourceFile("", code, ts.ScriptTarget.ESNext, false, ts.ScriptKind.TS);
  return Array.from(source.statements);
};

export const convertIntermediateCodes = (intermediateCodes: CodeGenerator.IntermediateCode[]): ts.Statement[] => {
  return intermediateCodes.reduce<ts.Statement[]>((result, intermediateCode) => {
    if (typeof intermediateCode === "string") {
      return [...result, ...stringToStatements(intermediateCode)];
    }
    return result.concat(intermediateCode);
  }, []);
};
