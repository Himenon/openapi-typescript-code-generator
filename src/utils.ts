export const parameterName = (operationId: string): string => `Parameter$${operationId}`;
export const requestBodyName = (operationId: string): string => `RequestBody$${operationId}`;
export const argumentParamsTypeDeclaration = (operationId: string): string => `Params$${operationId}`;
export const responseName = (operationId: string, statusCode: string): string => `Response$${operationId}$Status$${statusCode}`;
export const requestContentType = (operationId: string): string => `RequestContentType$${operationId}`;
export const responseContentType = (operationId: string): string => `ResponseContentType$${operationId}`;

export const isAvailableVariableName = (text: string): boolean => {
  return /^[A-Za-z_0-9\s]+$/.test(text);
};

export const escapeText = (text: string): string => {
  if (isAvailableVariableName(text)) {
    return text;
  }
  return `"${text}"`;
};

/** TODO escapeTextにマージする */
export const escapeText2 = (text: string): { escaped: boolean; text: string } => {
  if (isAvailableVariableName(text)) {
    return {
      escaped: false,
      text: text,
    };
  }
  return {
    escaped: true,
    text: `"${text}"`,
  };
};
