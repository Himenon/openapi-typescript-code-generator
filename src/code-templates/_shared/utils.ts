import ts from "typescript";

import type { TsGenerator } from "../../api";
import * as Utils from "../../utils";

export interface StringItem {
  type: "string";
  value: string;
}

export interface ExpressionItem {
  type: "property";
  value: ts.Expression;
}

export type Item = StringItem | ExpressionItem;

export type Params$TemplateExpression = Item[];

const getTemplateSpan = (
  factory: TsGenerator.Factory.Type,
  currentIndex: number,
  nextIndex: number,
  lastIndex: number,
  currentItem: ExpressionItem,
  nextItem: Item | undefined,
): ts.TemplateSpan[] => {
  // == last
  if (!nextItem) {
    return [
      factory.TemplateSpan.create({
        expression: currentItem.value,
        literal: factory.TemplateTail.create({
          text: "",
        }),
      }),
    ];
  }

  if (nextItem.type === "property") {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateMiddle.create({
            text: "",
          }),
        }),
      ];
    }
  } else {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: nextItem.value,
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateMiddle.create({
            text: nextItem.value,
          }),
        }),
      ];
    }
  }
};

/**
 *
 * ``
 * `a`
 * `${b}`
 * `a${b}`
 * `${a}${b}`
 * `${a}b${c}`
 * ``
 */
export const generateTemplateExpression = (factory: TsGenerator.Factory.Type, list: Params$TemplateExpression): ts.Expression => {
  if (list.length === 0) {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: "",
    });
  }
  const templateHead = factory.TemplateHead.create({
    text: list[0].type === "property" ? "" : list[0].value,
  });

  const spanList = list.splice(1, list.length);
  if (spanList.length === 0 && list[0].type === "string") {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: list[0].value,
    });
  }
  const lastIndex = spanList.length - 1;
  const restValue = lastIndex % 2;
  let templateSpans: ts.TemplateSpan[] = [];
  for (let i = 0; i <= (lastIndex - restValue) / 2; i++) {
    if (spanList.length === 0) {
      continue;
    }
    const currentIndex = 2 * i;
    const nextIndex = 2 * i + 1;
    const currentItem = spanList[currentIndex];
    const nextItem = spanList[nextIndex];
    if (currentItem.type === "string") {
      throw new Error("Logic Error");
    }
    templateSpans = templateSpans.concat(getTemplateSpan(factory, currentIndex, nextIndex, lastIndex, currentItem, nextItem));
  }
  return factory.TemplateExpression.create({
    head: templateHead,
    templateSpans: templateSpans,
  });
};

export interface VariableAccessIdentifer {
  kind: "string";
  value: string;
}

export interface ElementAccessExpression {
  kind: "element-access";
  value: string;
}

export type VariableElement = VariableAccessIdentifer | ElementAccessExpression;

export const splitVariableText = (text: string): VariableElement[] => {
  // ["...."] にマッチする
  const pattern = '["[a-zA-Z_0-9.]+"]';
  // 'a.b.c["a"]["b"]'.split(/(\["[a-zA-Z_0-9\.]+"\])/g)
  const splitTexts = text.split(/(\["[a-zA-Z_0-9\\.]+"\])/g); // 区切り文字も含めて分割
  return splitTexts.reduce<VariableElement[]>((splitList, value) => {
    if (value === "") {
      return splitList;
    }
    // ["book.name"] にマッチするか
    if (new RegExp(pattern).test(value)) {
      // ["book.name"] から book.name を抽出
      const matchedValue = value.match(/[a-zA-Z_0-9\\.]+/);
      if (matchedValue) {
        splitList.push({
          kind: "element-access",
          value: matchedValue[0],
        });
      }
      return splitList;
    } else {
      const dotSplited = value.split(".");
      const items = dotSplited.map<VariableAccessIdentifer>(childValue => ({ kind: "string", value: childValue }));
      return splitList.concat(items);
    }
  }, []);
};

export const generateVariableIdentifier = (
  factory: TsGenerator.Factory.Type,
  name: string,
): ts.Identifier | ts.PropertyAccessExpression | ts.ElementAccessExpression => {
  if (name.startsWith("/")) {
    throw new Error("can't start '/'. name=" + name);
  }
  const list = splitVariableText(name);
  // Object参照していない変数名の場合
  if (list.length === 1) {
    return factory.Identifier.create({
      name: name,
    });
  }
  const [n1, n2, ...rest] = list;
  // a.b のような単純な変数名の場合
  const first = factory.PropertyAccessExpression.create({
    expression: n1.value,
    name: n2.value,
  });

  return rest.reduce<ts.PropertyAccessExpression | ts.ElementAccessExpression>((previous, current) => {
    if (current.kind === "string" && Utils.isAvailableVariableName(current.value)) {
      return factory.PropertyAccessExpression.create({
        expression: previous,
        name: current.value,
      });
    }
    // 直接 .value でアクセスできない場合に ["value"] といった形で参照する
    return factory.ElementAccessExpression.create({
      expression: previous,
      index: current.value,
    });
  }, first);
};

export interface LiteralExpressionObject {
  [key: string]: { type: "string" | "variable"; value: string };
}

export const generateObjectLiteralExpression = (
  factory: TsGenerator.Factory.Type,
  obj: LiteralExpressionObject,
  extraProperties: ts.PropertyAssignment[] = [],
): ts.ObjectLiteralExpression => {
  const properties = Object.entries(obj).map(([key, item]) => {
    const initializer =
      item.type === "variable" ? generateVariableIdentifier(factory, item.value) : factory.StringLiteral.create({ text: item.value });
    return factory.PropertyAssignment.create({
      name: Utils.escapeText(key),
      initializer,
    });
  });

  return factory.ObjectLiteralExpression.create({
    properties: extraProperties.concat(properties),
    multiLine: true,
  });
};

/**
 * "/{a}/b/{a}/c{a}/".split(new RegExp("({a})"))
 * => ["/", "{a}", "/b/", "{a}", "/c", "{a}", "/"]
 */
export const stringToArray = (text: string, delimiter: string): string[] => text.split(new RegExp(`(${delimiter})`));

export const multiSplitStringToArray = (text: string, delimiters: string[]): string[] => {
  return delimiters.reduce<string[]>(
    (current, delimiter) => {
      return current.reduce<string[]>((result, currentText) => {
        return result.concat(stringToArray(currentText, delimiter));
      }, []);
    },
    [text],
  );
};
