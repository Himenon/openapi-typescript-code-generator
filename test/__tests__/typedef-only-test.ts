import * as fs from "fs";
import * as path from "path";

import * as Utils from "../utils";

describe("Typedef only", () => {
  test("typedef-api.test.domain", () => {
    const generateCode = fs.readFileSync(path.join(__dirname, "../code/typedef-only/api.test.domain.ts"), { encoding: "utf-8" });
    const text = Utils.replaceVersionInfo(generateCode);
    expect(text).toMatchSnapshot();
  });
  test("typedef-infer.domain", () => {
    const generateCode = fs.readFileSync(path.join(__dirname, "../code/typedef-only/infer.domain.ts"), { encoding: "utf-8" });
    const text = Utils.replaceVersionInfo(generateCode);
    expect(text).toMatchSnapshot();
  });
});
