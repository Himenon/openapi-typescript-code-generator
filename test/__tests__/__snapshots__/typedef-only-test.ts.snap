// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typedef only typedef-api.test.domain 1`] = `
"// 
// Generated by @himenon/openapi-typescript-code-generator
// 
// OpenApi : 3.1.0
// 
// License  : MIT
// 


export namespace Schemas {
    /** String Literal */
    export type StringType = string;
    export type StringHasEnumType = \\"a\\" | \\"A\\" | \\"b\\" | \\"B\\" | \\"c\\" | \\"C\\";
    export type StringDateType = string;
    export type StringDateTimeType = string;
    export type StringPasswordType = string;
    export type StringByteType = string;
    export type StringBinaryType = string;
    export type StringWithPatternType = string;
    /** Number Literal */
    export type NumberType = number;
    export type NumberHasEnumType = 1 | 2 | 3 | 100 | 123 | 0.1 | -0.1 | 0;
    export type NumberInt32Type = number;
    export type NumberInt64Type = number;
    export type NumberFloat = number;
    export type NumberDouble = number;
    /** Boolean Literal */
    export type BooleanType = boolean;
    export type ArrayStringType = string[];
    export type ArrayNumberType = number[];
    export type ArrayBooleanType = boolean[];
    export interface ObjectEmptyPropertiesType {
    }
    export interface ObjectHasPropertiesType {
        stringType?: string;
        numberType?: number;
        booleanType?: boolean;
        arrayType?: string;
        objectType?: {
            requiredStringType: string;
        };
    }
    export type OneOfType = string | number | {};
    export type AllOfType = {
        A?: string;
    } & {
        B?: string;
    };
    export type LocalRefStringType = Schemas.StringType;
    export type LocalRefNumberType = Schemas.NumberType;
    export type LocalRefBooleanType = Schemas.BooleanType;
    export type LocalRefArrayStringType = Schemas.BooleanType;
    export type LocalRefObjectHasPropertiesType = Schemas.ObjectHasPropertiesType;
    export interface LocalRefObjectProperties {
        stringType: Schemas.StringType;
        numberType: Schemas.NumberType;
        booleanType: Schemas.BooleanType;
        arrayType: Schemas.BooleanType;
        objectType: Schemas.ObjectHasPropertiesType;
    }
    export type LocalRefOneOfType = Schemas.StringType | Schemas.NumberType | Schemas.ObjectHasPropertiesType | Schemas.LocalRefObjectProperties;
    export type LocalRefAllOfType = Schemas.StringType & Schemas.NumberType & Schemas.ObjectHasPropertiesType & Schemas.LocalRefObjectProperties;
    export type LocalReferenceBeforeResolvedSchema1 = Schemas.UnresolvedTarget1;
    export type UnresolvedTarget1 = boolean;
    export type LocalReferenceBeforeResolvedSchema2 = Schemas.UnresolvedTarget2;
    export type UnresolvedTarget2 = Schemas.UnresolvedTarget3;
    export type UnresolvedTarget3 = number;
    export interface LocalReferenceBeforeResolvedSchema3 {
        unresolvedTarget4?: Schemas.UnresolvedTarget4;
    }
    export interface UnresolvedTarget4 {
        unresolvedTarget5?: Schemas.UnresolvedTarget5;
    }
    export type UnresolvedTarget5 = string;
    export type RemoteString = string;
    export type RemoteRefString = Schemas.RemoteString;
    export namespace Level1 {
        /** Level 1 */
        export type RemoteBoolean = boolean;
        export namespace Level2 {
            /** Level 2 */
            export type RemoteNumber = number;
            export namespace Level3 {
                /** Level 3 */
                export type RemoteArray = string[];
                export namespace Level4 {
                    /** Level 4 */
                    export interface RemoteObject {
                        A?: string;
                        B?: number;
                    }
                }
            }
        }
    }
    /** Level 1 */
    export type RemoteRefBoolean = Schemas.Level1.RemoteBoolean;
    /** Level 2 */
    export type RemoteRefNumber = Schemas.Level1.Level2.RemoteNumber;
    /** Level 3 */
    export type RemoteRefArray = Schemas.Level1.Level2.Level3.RemoteArray;
    /** Level 4 */
    export type RemoteRefObject = Schemas.Level1.Level2.Level3.Level4.RemoteObject;
    export namespace DirectRef {
        export type ForHeader = string;
        export interface ForResponse {
            forResponse?: string;
        }
        export type ForParameters = string;
        export interface ForRequestBody {
            a?: string;
        }
    }
    export namespace FullRemoteReference {
        export type ForParameters = string;
    }
    export interface Child {
        id: string;
        /** child name */
        name: string;
    }
    export interface Item {
        name: string;
        children: Child[];
    }
}
export namespace Headers {
    export type StringHeader = string;
    export type A = number;
    export namespace Level1 {
        export type B = string;
    }
    export type ReferenceOfHeaderToSchema = Schemas.DirectRef.ForHeader;
}
export namespace Responses {
    /**
     * Status Code 100
     * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/100
     */
    export namespace Continue { }
    /**
     * Status Code 101
     * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/101
     */
    export namespace SwitchingProtocol { }
    /** Type Reference - string */
    export namespace LocalReferenceStringDateTimeType {
        export interface Content {
            \\"application/json\\": Schemas.StringDateTimeType;
        }
    }
    /** Local Reference - object */
    export namespace LocalReferenceObjectType {
        export interface Content {
            \\"application/json\\": Schemas.ObjectHasPropertiesType;
        }
    }
    /** Response -> Schema */
    export namespace ReferenceOfResponseToSchema {
        export interface Content {
            \\"application/json\\": Schemas.DirectRef.ForResponse;
        }
    }
    /** response A definition */
    export namespace ResponseA {
        export interface Header {
            HeaderA: Headers.A;
            HeaderB: string;
        }
        export interface Content {
            \\"application/json\\": {
                /** responseA description */
                name?: \\"responseA\\";
            };
        }
    }
    export namespace Level1 {
        /** response B definition */
        export namespace ResponseB {
            export interface Content {
                \\"application/json\\": {
                    name?: \\"responseB\\";
                };
            }
        }
        export namespace Level2 {
            /** response C definition */
            export namespace ResponseC {
                export interface Content {
                    \\"application/json\\": {
                        name?: \\"responseC\\";
                    };
                }
            }
        }
    }
    export namespace ForPathItems {
        /** response A definition */
        export namespace FullRemoteReference {
            export interface Header {
                HeaderA: Headers.A;
                HeaderB: string;
            }
            export interface Content {
                \\"application/json\\": {
                    /** responseA description */
                    name?: \\"responseA\\";
                };
            }
        }
    }
}
export namespace Parameters {
    /** parameters.StringQueryParams */
    export type StringQueryParams = string;
    /** parameters.NumberQueryParams */
    export type NumberQueryParams = number;
    /** parameters.StringQueryParams */
    export type StringHeaderParams = string;
    export type A = number;
    export type RemoteReferenceA = Parameters.A;
    export namespace level1 {
        export type B = string;
    }
    export type RemoteReferenceB = Parameters.level1.B;
    /** parameters -> schemas */
    export type ReferenceOfParameterToSchema = Schemas.DirectRef.ForParameters;
    /** deepObject */
    export type DeepObjectParameter = {
        [key: string]: {
            gt?: string;
            gte?: string;
            lt?: string;
            lte?: string;
            any?: string | number | boolean;
        };
    };
}
export namespace RequestBodies {
    /** Request body string type */
    export namespace StringType {
        export interface Content {
            \\"application/json\\": string;
        }
    }
    /** Request body Local reference string type */
    export namespace LocalReferenceStringType {
        export interface Content {
            \\"application/json\\": Schemas.StringHasEnumType;
        }
    }
    /** requestBodies -> schemas */
    export namespace ReferenceOfRequestBodyToSchema {
        export interface Content {
            \\"application/json\\": Schemas.DirectRef.ForRequestBody;
        }
    }
    /** Remote Request body A */
    export namespace RequestBodyA {
        export interface Content {
            \\"application/json\\": {
                body?: string;
            };
        }
    }
    export namespace Level1 {
        /** Remote Request body B */
        export namespace RequestBodyB {
            export interface Content {
                \\"application/json\\": {
                    body?: string;
                };
            }
        }
        export namespace Level2 {
            /** Remote Request body C */
            export namespace RequestBodyC {
                export interface Content {
                    \\"application/json\\": {
                        body?: string;
                    };
                }
            }
        }
    }
}
"
`;

exports[`Typedef only typedef-infer.domain 1`] = `
"// 
// Generated by @himenon/openapi-typescript-code-generator
// 
// OpenApi : 3.1.0
// 
// License  : MIT
// 


export namespace Schemas {
    export type InferArrayEmptyItems = {}[];
    export type InferArrayType = string[];
    export interface InferObjectItem {
        description: any;
    }
    export interface InferObjectHasAnyProp {
        filed?: any;
    }
    export interface InferObject {
        filed?: string;
    }
    export type InferStringEnum = \\"a\\" | \\"b\\" | \\"c\\";
    export type InferAnyNullable = null;
}
"
`;
